modules/xfem/include/userobjects/MeshCut3DUserObject.h

```diff
@@ -0,0 +1,214 @@
+  virtual bool cutFragmentByGeometry(std::vector<std::vector<Point>> & frag_edges,
+                                     std::vector<CutEdge> & cut_edges,
+                                     Real time) const override;
+  virtual bool cutFragmentByGeometry(std::vector<std::vector<Point>> & frag_faces,
+                                     std::vector<CutFace> & cut_faces,
+                                     Real time) const override;
+
+protected:
+  /// The cutter mesh
+  std::unique_ptr<MeshBase> _cut_mesh;
+
+  /// The cutter mesh has triangluar elements only
+  const unsigned int _cut_elem_nnode = 3;
+  const unsigned int _cut_elem_dim = 2;
+
+  /// The structural mesh
+  MooseMesh & _mesh;
+
+  /// The structural mesh must be 3D only
+  const unsigned int _elem_dim = 3;
+
+  /// Used to define intersection points
+  const Real _const_intersection = 0.01;
+
+  /// Used for cutter mesh refinement and front advancement
+  Real _size_control;
+
+  /// Number of steps to grow the mesh
+  unsigned int _n_step_growth;
+
+  /// variables to help control the work flow
+  bool _stop;
+  bool _grow;
+
+  /// Boundary nodes of the cutter mesh
+  std::vector<dof_id_type> _boundary;
+
+  /// Edges at the boundary
+  std::set<CutEdge> _boundary_edges;
+
+  /// A map of boundary nodes and their neighbors
+  std::map<dof_id_type, std::vector<dof_id_type>> _boundary_map;
+
+  /// Active boundary nodes where growth is allowed
+  std::vector<std::vector<dof_id_type>> _active_boundary;
+
+  /// Growth direction for active boundaries
+  std::vector<std::vector<Point>> _active_direction;
+
+  /// Inactive boundary
+  std::vector<unsigned int> _inactive_boundary_pos;
+
+  /// New boundary after growth
+  std::vector<std::vector<dof_id_type>> _front;
+
+  /**
+    Check if a line intersects with an element
+   */
+  virtual bool intersectWithEdge(const Point & p1,
+                                 const Point & p2,
+                                 const std::vector<Point> & _vertices,
+                                 Point & pint) const;
+
+  /**
+    Find directional intersection along the positive extension of the vector from p1 to p2
+   */
+  bool findIntersection(const Point & p1,
+                        const Point & p2,
+                        const std::vector<Point> & vertices,
+                        Point & pint) const;
+
+  /**
+    Check if point p is inside the edge p1-p2
+   */
+  bool isInsideEdge(const Point & p1, const Point & p2, const Point & p) const;
+
+  /**
+    Get the relative position of p from p1
+   */
+  Real getRelativePosition(const Point & p1, const Point & p2, const Point & p) const;
+
+  /**
+    Check if point p is inside a plane
+   */
+  bool isInsideCutPlane(const std::vector<Point> & _vertices, Point p) const;
+
+  /**
+    Find the normal vector to an element
+   */
+  void findElemNormal(const std::vector<Point> & _vertices, Point & _normal, Point & _center) const;
```

Comment:
I feel like a lot of these geometry methods are re-inventing the wheel. Before we add another round of geometry code to the modules, it would be nice to review what might be overlapping. Take a look in libMesh at the `Plane`, `Line` and related objects. We can already do normals, projections, with basic types. It's in libMesh. On top of that, we've implemented additional search routines in our geometric search system. Projections of nodes onto planes, lines onto planes, intersections, etc. Now, I'm see it all again for this object. We have to get better about code reuse. I'm more than willing to help find this or assist you in finding these so we don't keep adding to the problem.

Reply:
Thanks for offering the help.  I found "Plane" and "ElemCutter" in libMesh quite useful, and I even started making a couple of changes.  Then I realized that all methods of making cuts and finding normals in this PR (MeshCut3DUserObject) were coded based on the existing code in GeometricCut3DUserObject. It seems to me that updating these geometric methods in both objects with the libMesh objects might be a separate Issue and PR.   Other than this, I have done changes to address the comments. Please advise if I should push what I have (and I would be glad to help with updating geometric methods later), or continue changes with the libMesh objects. Either is fine to me. @permcody @bwspenc @jiangwen84

Reply:
I have used "Plane.unit_normal" to calculate the normal to an element. This removes a method "findElemNormal", improving this PR. For other methods relating to cutting, I will not change in this PR because of the reason stated above. I will push what I have for your review. @permcody @bwspenc @jiangwen84